### PR TITLE
Add tooltip for labels

### PR DESCRIPTION
**Pull request by Czaki**
# Description

This PR adds tooltips for Labels and Points (partially) layer. For full work with Points layer there is a need to overwrite `get_status` for this class. 

This PR address  #2622 issue 

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by Czaki** [comment]
I do not understand fail of these tests.

**Comment by Carreau** [comment]
I believe the test error are random – at least one is, the dask one –  I'm going to try closing and reopening see if rerun the test works.

**Comment by goanpeca** [comment]
@Czaki could you add a screenshot of what this enables :) ?

Thanks!

**Comment by Czaki** [comment]
@goanpeca 
Zrzut ekranu z 2021-05-24 10-16-27

**Comment by goanpeca** [comment]
Thanks @Czaki 

That looks pretty cool. My question would be regarding the style of the QToolTip.

**Comment by Czaki** [comment]
> That looks pretty cool. My question would be regarding the style of the QToolTip.

I o not have experience with styling QToolTip. This is the default style. The text comes from `get_status` functions splitter on `,`.

File: napari/_qt/qt_main_window.py [hunk 1]
```diff
@@ -109,6 +110,10 @@ def event(self, e):
                 inst.append(inst.pop(inst.index(self)))
             except ValueError:
                 pass
+        if e.type() == QEvent.ToolTip and self.qt_viewer.viewer.tooltip_text:
+            QToolTip.showText(
+                e.globalPos(), self.qt_viewer.viewer.tooltip_text
+            )
```
Comment (goanpeca) [inline napari/_qt/qt_main_window.py]: I think it has been a Qt bug for a long time that using QToolTip directly does not use the stylesheet, so in order to preserve this one needs to use a custom widget that emulates a tooltip.

Comment (Czaki) [reply napari/_qt/qt_main_window.py]: I do not know about this bug. 

Here result after the last commit: 

Zrzut ekranu z 2021-05-24 16-48-18

Comment (goanpeca) [reply napari/_qt/qt_main_window.py]: Much better! thanks for the fix @Czaki

**Review by sofroniewn — CHANGES_REQUESTED** [review]
Hi @Czaki, as we discussed at the euro dev meeting a while back, I think this functionality should be strictly opt-in, and it should be easy for people to turn on/ off or customize.

One possible approach is that we mirror what we've done with [scale bar](https://github.com/napari/napari/blob/09fc5993d61999a26c092939c0dd56f054d6383a/napari/components/scale_bar.py#L8) and [axes](https://github.com/napari/napari/blob/09fc5993d61999a26c092939c0dd56f054d6383a/napari/components/axes.py#L4) and have a new

```python
from ..utils.events import EventedModel


class Tooltip(EventedModel):
    """Tooltip showing additional information on the cursor.

    Attributes
    ----------
    visible : bool
        If tooltip is visible or not.
    # OTHER PARAMETERS ......
    """

    # fields
    visible: bool = False
    # OTHER PARAMETERS ......
```

As to what other parameters we should allow / how we should let people customize I'm not sure. This was also something that @VolkerH requested a while back in #531, and others may have ideas on what they want to see here too

**Comment by Czaki** [comment]
Hi @sofroniewn. I still do not understand the idea of adding a new model. Controlling of visibility should rather live in settings. Or I'm wrong?

Or you think that the visibility of the +tooltip should be controlled for every layer separately? 

currently I use `get_message` output and it should be place for such callback live like in #531.

**Comment by sofroniewn** [comment]
> Hi @sofroniewn. I still do not understand the idea of adding a new model. Controlling of visibility should rather live in settings. Or I'm wrong?

We could move all of this information to our setting object, but that would start becoming very large. Also in that case `viewer.tooltip_text` as is added in this PR should not exist either.

I can see though how things like `viewer.axes` and `viewer.scale_bar` should be in settings too though, but that is a larger conversation than this PR.

Curious what @goanpeca thinks here.

> Or you think that the visibility of the +tooltip should be controlled for every layer separately?

No, the proposal that I've made doesn't do that either.

> currently I use get_message output and it should be place for such callback live like in #531.

I guess should the status bar and the tooltip always show the same thing? Should you be able to customize either of them/ both of them. It seems like it would be nice to be able to customize both of them separately, and there has been some discussion of this elsewhere.

**Comment by Czaki** [comment]
I add a separate event tooltip model (with default tooltip enabled) and add the `get_tooltip_text` method to the base layer class.

**Comment by Czaki** [comment]
The current view:
 
Zrzut ekranu z 2021-06-01 00-18-24

**Comment by goanpeca** [comment]
>As to what other parameters we should allow / how we should let people customize I'm not sure. This was also something that @VolkerH requested a while back in #531, and others may have ideas on what they want to see here too

Should we add a setting for this then ?

**Comment by goanpeca** [comment]
In a subsequent PR I will try to connect 2-way settings to model options, so that things work as expected, changing the settings changes the models and viceversa.

**Review by goanpeca — APPROVED** [review]
Thanks @Czaki

**Comment by Czaki** [comment]
@goanpeca I just added settings and synchronization.

**Comment by goanpeca** [comment]
>@goanpeca I just added settings and synchronization.

Thanks. Testing

**Comment by goanpeca** [comment]
This now has a merge conflict @Czaki 

Also maybe ready for another review round @sofroniewn

**Comment by jni** [comment]
I think it makes more sense to go with @sofroniewn's original proposal of having tooltip visibility controlled by the menubar menu, together with axes and scale bar. A setting can control the default *on launch*, but in a session it should be controlled by the menu rather than by a setting.

I also would prefer to have this off by default.

**Comment by sofroniewn** [comment]
> I think it makes more sense to go with @sofroniewn's original proposal of having tooltip visibility controlled by the menubar menu, together with axes and scale bar. A setting can control the default on launch, but in a session it should be controlled by the menu rather than by a setting.

Yup agreed

> I also would prefer to have this off by default.

also agreed

I'm also wondering if we want to add a `get_tooltip_text` to the base layer, I feel like the layer should be able to tell you something about it's status/ properties, but then it's the responsibility of something else to convert that/ process that into the tooltip, and it will could be configured what goes in depending on preferences and settings, but that doesn't need to leak down into every layer. Is there a way you can do this PR without adding that new method to the base layer? (which only returns a non-empty string for labels).

**Comment by jni** [comment]
> I'm also wondering if we want to add a get_tooltip_text to the base layer

Agreed on this also. Possibly the happy path is to tackle this after unifying properties, but we can also make it a private method on labels for now to get going, then unify.

**Comment by sofroniewn** [comment]
> private method on labels for now to get going, then unify.

this would be ok for now too

**Comment by Czaki** [comment]
It is possible to add something to settings without rendering it inside the settings dialog? Or it is ok to be switchable in menubar and settings dialog?

**Comment by Czaki** [comment]
I have hope that I address all remarks.

File: napari/components/viewer_model.py [hunk 2]
```diff
@@ -368,6 +378,9 @@ def _on_cursor_position_change(self, event):
         if active is not None:
             self.status = active.get_status(self.cursor.position, world=True)
             self.help = active.help
+            self.tooltip.text = active._get_tooltip_text(
```
Comment (sofroniewn) [inline napari/components/viewer_model.py]: Can we put this inside `if self.tooltip.visibile:` so that it doesn't get called otherwise

Comment (Czaki) [reply napari/components/viewer_model.py]: done

**Review by sofroniewn — APPROVED** [review]
Thanks for those changes, this now looks good to me!

**Review by jni — APPROVED** [review]

**Pull request merged**